cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -80,35 +90,87 @@ public RemoteTaskExecutor(CConfiguration cConf, RemoteClientFactory remoteClient
+          if (compression) {
+            requestBuilder.addHeader(HttpHeaders.CONTENT_ENCODING, "gzip");
+            requestBuilder.addHeader(HttpHeaders.ACCEPT_ENCODING, "gzip, deflate");
+          }
+          HttpRequest httpRequest = requestBuilder.build();
 
-        HttpResponse httpResponse = remoteClient.execute(httpRequest);
-        if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+          HttpResponse httpResponse = remoteClient.execute(httpRequest);
+          if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+            throw new RetryableException(
+              String.format("Received response code %s for %s", httpResponse.getResponseCode(),
+                            runnableTaskRequest.getClassName()));
+          }
+          if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+            BasicThrowable basicThrowable = GSON
+              .fromJson(new String(getResponseBody(httpResponse)), BasicThrowable.class);
+            throw RemoteExecutionException.fromBasicThrowable(basicThrowable);
+          }
+          return new RetryableTaskResult(getResponseBody(httpResponse), attempt);
+        } catch (NoRouteToHostException e) {
           throw new RetryableException(
-            String.format("Received response code %s for %s", httpResponse.getResponseCode(),
-                          runnableTaskRequest.getClassName()));
-        }
-        if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
-          BasicThrowable basicThrowable = GSON
-            .fromJson(new String(getResponseBody(httpResponse)), BasicThrowable.class);
-          throw RemoteExecutionException.fromBasicThrowable(basicThrowable);
+            String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
         }
-        return getResponseBody(httpResponse);
-      } catch (NoRouteToHostException e) {
-        throw new RetryableException(
-          String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
+      }, retryStrategy, Retries.DEFAULT_PREDICATE);
+      //emit metrics with successful result
+      emitMetrics(startTime, true, runnableTaskRequest, retryableTaskResult.getAttempts());
+      return retryableTaskResult.getResult();
+    } catch (Exception e) {
+      //emit metrics with failed result
+      emitMetrics(startTime, false, runnableTaskRequest, getAttempts(e));
+      throw e;
+    }
+  }
+
+  /**
+   * Find if attempt count is included in the suppressed throwable
+   *
+   * @param e Exception to analyze
+   * @return attempt count
+   */
+  private int getAttempts(Exception e) {
+    int attempts = 0;
+    Throwable[] suppressed = e.getSuppressed();
+    for (Throwable t : suppressed) {
+      if (t instanceof RetryFailedException) {
+        attempts = ((RetryFailedException) t).getRetries();
```

Comment:
Better define a meaningful interface to expose the `getRetries()` method instead of scattering into different classes.

Reply:
Done